xla/service/collective_ops_utils.cc

```
@@ -779,6 +809,34 @@ bool IsCollective(const HloInstruction* instruction) {
   return false;
 }
 
+bool IsAsyncCollective(const HloInstruction* instruction) {
+  if (instruction->IsAsynchronous()) {
+    switch (instruction->async_wrapped_opcode()) {
+      case HloOpcode::kAllGather:
+      case HloOpcode::kAllReduce:
+      case HloOpcode::kAllToAll:
+      case HloOpcode::kCollectiveBroadcast:
+      case HloOpcode::kCollectivePermute:
+      case HloOpcode::kRaggedAllToAll:
+      case HloOpcode::kReduceScatter:
+        return true;
+      default:
+        return false;
```

Can you list the false cases too and throw an error in the default case? I'm almost sure you're missing collectives here, e.g. send/recv

---

updated